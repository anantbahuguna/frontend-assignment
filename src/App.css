*,
*::after,
*::before {
  box-sizing: border-box;
  margin: 0;
  padding: 0;
}

:root {
  /* variables */
}

body {
  font-family: sans-serif;
}

.container {
  /* display: flex;
  flex-direction: column; */
  margin: 10px;
  /* align-items: center; */
  text-align: center;
  width: 400px;
  max-width: 90vw;
}

.bottom-container {
  margin: 10px;
}
