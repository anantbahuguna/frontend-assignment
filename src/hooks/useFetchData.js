import { useEffect, useState } from "react";

// In production, we would use some library like react-query which gives us features like caching, refetching, etc.
export default function useFetchData(url) {
  const [data, setData] = useState([]);
  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useState(null);

  useEffect(() => {
    if (!url) {
      return;
    }
    let ignore = false;
    const fetchData = async () => {
      setIsLoading(true);
      try {
        const response = await fetch(url);
        if (!response.ok) {
          throw new Error(`Response status: ${response.status}`);
        }
        const newData = await response.json();
        if (!ignore) {
          setData(newData);
        }
      } catch (err) {
        if (err instanceof Error) {
          setError(err.message);
        } else {
          console.error("Failed to fetch data: ", err);
          throw err;
        }
      } finally {
        setIsLoading(false);
      }
    };

    fetchData();

    return () => {
      ignore = true;
    };
  }, [url]);
  return { data, isLoading, error };
}
