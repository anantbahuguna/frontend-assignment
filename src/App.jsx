import { useMemo, useState } from "react";
import Pagination from "./components/Pagination/Pagination";
import ProjectsTable from "./components/ProjectsTable/ProjectsTable";
import Loader from "./components/Loader/Loader";
import useFetchData from "./hooks/useFetchData";
import "./App.css";

const API_URL =
  "https://raw.githubusercontent.com/saaslabsco/frontend-assignment/refs/heads/master/frontend-assignment.json";
const ITEMS_PER_PAGE = 5;

export default function App() {
  const { data, isLoading, error } = useFetchData(API_URL);

  const [currentPage, setCurrentPage] = useState(0);
  const [itemsPerPage, setItemsPerPage] = useState(ITEMS_PER_PAGE);

  const currentData = useMemo(
    () =>
      data?.slice(currentPage * itemsPerPage, (currentPage + 1) * itemsPerPage),
    [data, itemsPerPage, currentPage]
  );

  if (error) {
    return <h2>Something went wrong! Please try again.</h2>;
  }

  return (
    <div className="app">
      {isLoading ? (
        <Loader />
      ) : (
        data && (
          <div className="container">
            <ProjectsTable projects={currentData} />
            <Pagination
              totalItems={data.length}
              itemsPerPage={itemsPerPage}
              currentPage={currentPage}
              onPageChange={setCurrentPage}
              setItemsPerPage={setItemsPerPage}
              setCurrentPage={setCurrentPage}
              visibleRange={1}
            />
          </div>
        )
      )}
    </div>
  );
}
