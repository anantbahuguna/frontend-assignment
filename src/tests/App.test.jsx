import React from "react";
import { render, screen, waitFor } from "@testing-library/react";
import App from "../App"; // Adjust the path as necessary
import { mockProjectsData } from "./mocks/mockProjectsData";
import useFetchData from "../hooks/useFetchData";

jest.mock("../hooks/useFetchData");

describe("App Component", () => {
  test("matches snapshot for loading state", () => {
    useFetchData.mockReturnValue({
      data: [],
      isLoading: true,
      error: null,
    });

    const { asFragment } = render(<App />);
    expect(asFragment()).toMatchSnapshot();
  });

  test("renders App with fetched data", async () => {
    useFetchData.mockReturnValue({
      data: mockProjectsData,
      isLoading: false,
      error: null,
    });

    render(<App />);

    expect(screen.getByText("15823")).toBeInTheDocument();
    expect(screen.getByText("6859")).toBeInTheDocument();
  });

  test("renders error message when fetch fails", () => {
    useFetchData.mockReturnValue({
      data: [],
      loading: false,
      error: "Failed to fetch",
    });

    render(<App />);
    expect(
      screen.getByText("Failed to fetch", { exact: false })
    ).toBeInTheDocument();
  });
});
