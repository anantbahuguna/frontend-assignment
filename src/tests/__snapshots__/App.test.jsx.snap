// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component matches snapshot for loading state 1`] = `
<DocumentFragment>
  <div
    class="app"
  >
    <div
      aria-label="Loading"
      class="loader"
      data-test-id="loader"
    />
  </div>
</DocumentFragment>
`;
